File: devlibrary/src/main/java/com/example/devlibrary/rxbus/RxBus.java
package com.example.devlibrary.rxbus;

import io.reactivex.Flowable;
import io.reactivex.processors.FlowableProcessor;
import io.reactivex.processors.PublishProcessor;
import io.reactivex.subscribers.SerializedSubscriber;

public enum RxBus {
    INSTANCE;

    private final FlowableProcessor<Object> bus;

    // PublishSubject只会把在订阅发生的时间点之后来自原始Observable的数据发射给观察者
    RxBus() {
        bus = PublishProcessor.create().toSerialized();
    }

    // 发送一个新的事件
    public void post(Object o) {
        new SerializedSubscriber<>(bus).onNext(o);
    }

    // 根据传递的 eventType 类型返回特定类型(eventType)的 被观察者
    public <T> Flowable<T> toFlowable(Class<T> eventType) {
        return bus.ofType(eventType);
    }

    public boolean hasSubscribers() {
        return bus.hasSubscribers();
    }
}

File: baselib/src/main/java/com/wan/baselib/rxbus/RxBus.kt
package com.wan.baselib.rxbus

import io.reactivex.rxjava3.core.Flowable
import io.reactivex.rxjava3.processors.FlowableProcessor
import io.reactivex.rxjava3.processors.PublishProcessor
import io.reactivex.rxjava3.subscribers.SerializedSubscriber

enum class RxBus {
    INSTANCE;

    private val bus: FlowableProcessor<Any> = PublishProcessor.create<Any>().toSerialized()

    // 发送一个新的事件
    fun post(o: Any) {
        SerializedSubscriber(bus).onNext(o)
    }

    // 根据传递的 eventType 类型返回特定类型(eventType)的 被观察者
    fun <T : Any> toFlowable(eventType: Class<T>): Flowable<T> {
        return bus.ofType(eventType)
    }

    fun hasSubscribers(): Boolean {
        return bus.hasSubscribers()
    }

}